xdp/src/route_monitor.rs

```diff
@@ -0,0 +1,362 @@
+use {
+    crate::{
+        netlink::{
+            is_supported_ipv4_neigh_header, is_supported_ipv4_route_header, is_valid_link_update,
+            parse_ifinfomsg, parse_rtm_newneigh, parse_rtm_newroute, NetlinkMessage, NetlinkSocket,
+        },
+        route::{AtomicRouter, WorkingRouter},
+    },
+    libc::{
+        poll, pollfd, POLLERR, POLLHUP, POLLIN, POLLNVAL, RTMGRP_IPV4_ROUTE, RTMGRP_LINK,
+        RTMGRP_NEIGH, RTM_DELLINK, RTM_DELNEIGH, RTM_DELROUTE, RTM_NEWLINK, RTM_NEWNEIGH,
+        RTM_NEWROUTE,
+    },
+    log::*,
+    std::{
+        io::{Error, ErrorKind},
+        net::IpAddr,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread,
+        time::{Duration, Instant},
+    },
+};
+pub struct RouteMonitor;
+
+impl RouteMonitor {
+    /// subscribes RTMGRP_IPV4_ROUTE | RTMGRP_NEIGH and drain for `drain_window`
+    /// if any relevant updates arrived, update working routing table
+    /// once drain window expires, publish the working routing table to the atomic router
+    pub fn start(
+        atomic_router: Arc<AtomicRouter>,
+        exit: Arc<AtomicBool>,
+        drain_window: Duration,
+    ) -> thread::JoinHandle<()> {
+        thread::spawn(move || {
+            let groups = (RTMGRP_IPV4_ROUTE | RTMGRP_NEIGH | RTMGRP_LINK) as u32;
+            let mut sock = match NetlinkSocket::open_multicast_listener(groups) {
+                Ok(s) => s,
+                Err(e) => {
+                    error!("netlink bind failed: {e}");
+                    return;
+                }
+            };
+
+            let mut working = None;
+            loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+                if working.is_none() {
+                    working = Some(WorkingRouter::from_atomic_router(&atomic_router));
+                }
+
+                let mut pfd = pollfd {
+                    fd: sock.as_raw_fd(),
+                    events: POLLIN,
+                    revents: 0,
+                };
+                // block until data is available
+                let rv = match poll_helper(&mut pfd, -1) {
+                    Ok(rv) => rv,
+                    Err(e) => {
+                        warn!("poll error: {e}");
+                        let _ = atomic_router.resync();
+                        working = None;
+                        continue;
+                    }
+                };
+
+                // handle poll event and check for errors
+                match handle_revents(rv, pfd.revents, &atomic_router, &mut working) {
```

Comment:
I'm confused by this poll return => Revents conversion => LoopControl conversion

IMO this is making things more not less complicated, and moving logic away from
where it should be? For example we panic in handle_revents on Revents::Nval,
which is just POLLNVAL, and even the log calls it POLLNVAL? Can't we panic
without converting to Revents? 😅

I would remove all these abstraction layers, and I'm also prettty sure that we
can do this loop without polling/calling handle_revents twice?